feat(blog): add Yarn workspaces for theme development blog post (#12996)
gatsbyjs/gatsby: docs/blog/2019-04-01-setting-up-yarn-workspaces-for-theme-development/index.md

---
title: Setting up Yarn Workspaces for Theme Development
date: 2019-04-01
author: Brent Jackson
excerpt: ""
tags:
  - themes
  - yarn
  - tutorials
---

Gatsby [themes][themes-tag] are a new way to share functionality across multiple Gatsby sites.
Using [Yarn workspaces][] is an excellent way to set up a project for theme development.
It allows you to keep multiple packages in a single parent directory and link dependencies together.
For Gatsby theme development, that means you can keep multiple themes and example sites together in a single project.
In this post, we'll walk through how to use Yarn workspaces to set up a development environment for creating custom themes.

## Installation and directory structure

First, if you don't have Yarn installed already, follow the directions here to [install Yarn][].
Next, create a new directory for the entire project, where you'll be adding your theme and an example site later.

Create two subdirectories in this folder: one for the theme itself and one for the example site.

```shell
mkdir gatsby-theme-example-workspaces example

> ```

Author: Yeah, great idea! I'll add one in. Do you think it makes sense to go here or after the `yarn init` step?

> ```

Reviewer: It would work in either place, I think it's fine where you have it. 👍

Add a `package.json` to the root directory with these subdirectories in the `workspaces` field. Note that setting `"private": true` is required for Yarn workspaces.

```json:title=package.json
{
  "private": true,
  "workspaces": ["gatsby-theme-example-workspaces", "example"]
}
```

> Switch to each subdirectory and run `yarn init` to create a `package.json` for each one.

Author: Maybe it’s fixed now, but ran into issues where shadowing wouldn’t work if the theme’s folder name was different. I’ll give it a shot and remove that language if it seems I work

Be sure the `name` field in your theme's `package.json` matches the directory name exactly.

From the root directory, run the following to install Gatsby's dependencies for the example site.

```shell
yarn workspace example add gatsby react react-dom
```

The `yarn workspace` command will run Yarn commands for a specific workspace without needing to switch directories.

> Also add the following as peer dependencies to the theme.

Reviewer 2: ```suggestion
Then add the following as peer dependencies to the theme.
```

> This will help ensure that the end user of your theme can choose to use any compatible version of Gatsby with your theme.

Reviewer 2: ```suggestion
This will ensure that the end user of your theme can choose any compatible version of Gatsby.
```

```shell
yarn workspace gatsby-theme-example-workspaces add --peer gatsby react react-dom
```

## Base theme setup

To ensure that Gatsby will compile ES6 syntax in your theme, add the following plugin to its `gatsby-config.js`.
Note that this is only needed temporarily. Gatsby will automatically transpile themes in a later version.

```shell
yarn workspace gatsby-theme-example-workspaces add gatsby-plugin-compile-es6-packages
```

```js:title=gatsby-theme-example-workspaces/gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: "gatsby-plugin-compile-es6-packages",
      options: {
        modules: ["gatsby-theme-example-workspaces"],
      },
    },
  ],
}
```

> Create an empty `index.js` file to the theme's directory that serves as a noop (empty function).

Author: Yeah, I don't have a strong opinion here – was mostly going off how this is described elsewhere. Pointing to the `gatsby-config.js` seems a *little* odd to me because it doesn't seem like a module that end users would need to import very often. But if this simplifies the steps in this post, then I think it makes sense

---
title: Setting up Yarn Workspaces for Theme Development
date: 2019-04-01
author: Brent Jackson
excerpt: ""
tags:
  - themes
  - yarn
  - tutorials
---

Gatsby [themes][themes-tag] are a new way to share functionality across multiple Gatsby sites.
Using [Yarn workspaces][] is an excellent way to set up a project for theme development.

> It allows you to keep multiple packages in a single parent directory and link dependencies together.

Reviewer: ```suggestion
[Yarn workspaces][] are an excellent way to set up a project for theme development because they allow you to keep multiple packages in a single parent directory and link dependencies together.
```
This and the previous sentence might read a little better if combined.

For Gatsby theme development, that means you can keep multiple themes and example sites together in a single project.
In this post, we'll walk through how to use Yarn workspaces to set up a development environment for creating custom themes.

## Installation and directory structure

First, if you don't have Yarn installed already, follow the directions here to [install Yarn][].
Next, create a new directory for the entire project, where you'll be adding your theme and an example site later.

Create two subdirectories in this folder: one for the theme itself and one for the example site.

```shell
mkdir gatsby-theme-example-workspaces example
```

Add a `package.json` to the root directory with these subdirectories in the `workspaces` field. Note that setting `"private": true` is required for Yarn workspaces.

```json:title=package.json
{
  "private": true,
  "workspaces": ["gatsby-theme-example-workspaces", "example"]
}
```

> Switch to each subdirectory and run `yarn init -y` to create a `package.json` for each one.

Reviewer: ```suggestion
Change to each subdirectory and run `yarn init -y` to create a `package.json` for each one.
```
`Change` might signal the command-line workflow a little more than `Switch`.

Be sure the `name` field in your theme's `package.json` matches the directory name exactly.

> This is currently a limitation of the shadowing feature, not Yarn workspaces.

Reviewer 3: pls explain also: which shadowing feature?

Reviewer: It could say `This is currently a limitation of Gatsby theme shadowing` instead?

module.exports = {
  plugins: [
    {
      resolve: "gatsby-plugin-compile-es6-packages",
      options: {
        modules: ["gatsby-theme-example-workspaces"],
      },
    },
  ],
}
```

Change the `main` field in your theme's `package.json` to point to the `gatsby-config.js` file.

```json:title=gatsby-theme-example-workspaces/package.json
{
  "name": "gatsby-theme-example-workspaces",
  "description": "",
  "version": "1.0.0",
  "main": "gatsby-config.js", // highlight-line
  "license": "MIT",
  "peerDependencies": {
    "gatsby": "^2.2.10",
    "react": "^16.8.5",
    "react-dom": "^16.8.5"
  },
  "dependencies": {
    "gatsby-plugin-compile-es6-packages": "^1.0.6"
  }
}
```

## Example site setup

Add the theme as a dependency to the example site.
By specifying the version here, Yarn will install the local dependency that hasn't been published to npm yet.

```shell
yarn workspace example add gatsby-theme-example-workspaces@1.0.0
```

In the example site, create a `gatsby-config.js` file and add the theme.

```js:title=example/gatsby-config.js
module.exports = {
  __experimentalThemes: ["gatsby-theme-example-workspaces"],
}
```

> Add a `src/pages/` directory and add a simple _Hello, world_ page.

Reviewer: ```suggestion
Add a `src/pages/` directory and add a _Hello, world_ page.
```

```js:title=example/src/pages/index.js
import React from "react"

export default props => <h1>Hello, world</h1>
```

> Add Gatsby scripts to the example site's `package.json`.

Reviewer: ```suggestion
Add Gatsby develop and build scripts to the example site's `package.json`.
```

```shell
yarn workspace gatsby-theme-example-workspaces add gatsby-plugin-page-creator
```

```js:title=gatsby-theme-example-workspaces/gatsby-config.js
const path = require("path") // highlight-line
module.exports = {
  plugins: [
    // highlight-start
    {
      resolve: "gatsby-plugin-page-creator",
      options: {
        path: path.join(__dirname, "src/pages"),
      },
    },
    // highlight-end
    {
      resolve: "gatsby-plugin-compile-es6-packages",
      options: {
        modules: ["gatsby-theme-example-workspaces"],
      },
    },
  ],
}
```

Make a `src/pages` directory in the theme and add a demo page.

```js:title=gatsby-theme-example-workspaces/src/pages/theme-page.js
import React from "react"

export default props => <h1>Hello, from the theme!</h1>
```

Stop and restart the Gatsby development server to pick up the new page from the theme. The theme's page should be visible at `http://localhost:8080/theme-page`.

That's it! By now you should have a basic Yarn workspaces setup to develop Gatsby themes with.
Be sure to look for more posts on developing Gatsby themes in the near future,

> and you can read more about themes on the [blog][themes-tag].

Reviewer: ```suggestion
and you can read more about themes here on the [blog][themes-tag].
```
